### DISCUSSION TITLE
Caluclating the spatial derivative of a material peroperty

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#25578] [Q&A General]
Hello,

I wanted to write a kernel to calculate the derivative of a material property w.r.t x, y and z. I just do not know how to declare them in the .h and .C files. Is there any specific example that I can look through?

Thanks,
Ali

**Comment by GiudGiud** [comment]
Hello

Do you want to declare each derivative separately?
Or all of them in a vector (gradient) ?

Guillaume

**Reply by alimostafavi24** [reply]
I need them separately (by component).

Can they also be used for determining the residual and off diag jacobian?

**Reply by GiudGiud** [reply]
you can just declare 3 properties then, changing the name using a utility as can be seen in DerivativeParsedMaterial.C for example

You ll need to write the expression yourself.
for example if M = u (variable), then you ll need to compute dM/dx = du/dx and for that you ll retrieve coupledGrdient(u) and retrieve it at the right index
